File: 345_reverse_vowels_of_a_string/345_reverse_vowels_of_a_string.py
#!/usr/bin/env python
#-*- coding: UTF-8 -*-

##########################################################################
#	> File Name: 345_reverse_vowels_of_a_string.py
#	> Author: Tingjian Lau
#	> Mail: tjliu@mail.ustc.edu.cn
#	> Created Time: 2016/05/11
#########################################################################

# Time: O(n)
# Space: O(1)
def reverseString_1(s):
    vowels = 'aeiou'
    string = list(s)
    p, q = 0, len(s)-1

    while(p < q):
        if string[p].lower() not in vowels:
            p += 1
        elif string[q].lower() not in vowels:
            q -= 1;
        else:
            string[p], string[q] = string[q], string[p]
            p += 1
            q -= 1

    return "".join(string)

if __name__ == '__main__':
    print reverseString_1('leetcode')
    print reverseString_1('hello')
    print reverseString_1('aeiou')
    print reverseString_1('')

File: 345_reverse_vowels_of_a_string/345_reverse_vowels_of_a_string.cpp
/*************************************************************************
	> File Name: 345_reverse_vowels_of_a_string.cpp
	> Author: Tingjian Lau
	> Mail: tjliu@mail.ustc.edu.cn
	> Created Time: 2016/05/11
 ************************************************************************/

#include <iostream>
#include <string>

using namespace std;

bool isVowel(char ch){
    const string vowels = "aeiou";
    
    return vowels.find(tolower(ch)) != string::npos;
}

// Time: O(n)
// Space: O(1)
string reverseVowels_1(string s){
    for(int p=0, q=s.length()-1; p<q;){
        if(!isVowel(s[p]))
            ++p;
        else if(!isVowel(s[q]))
            --q;
        else{
            swap(s[p++], s[q--]);
        }
    }

    return s;
}

int main(){
    cout << reverseVowels_1("leetcode") << endl;
    cout << reverseVowels_1("hello") << endl;
    cout << reverseVowels_1("aeiou") << endl;
    cout << reverseVowels_1("") << endl;

    return 0;
}
